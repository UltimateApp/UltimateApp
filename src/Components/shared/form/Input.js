import React, { useState } from 'react';
import { StyleSheet, TextInput, View, Text } from 'react-native';
import { useField } from 'formik';

import theme from '../../../styles/theme.style';

const Input = ({ fieldName, label, ...props }) => {
  const [isFocused, setIsFocused] = useState(false);
  const [field, meta] = useField(fieldName);

  const borderColor = isFocused ? styles.focusedInput : meta.error ? styles.errorInput : undefined;
  return (
    <View style={styles.group}>
      <Text style={styles.label}>{label}</Text>
      <View style={[styles.input, borderColor]}>
        {meta.error && meta.touched && <Text style={styles.error}>{meta.error}</Text>}
        <TextInput
          onChangeText={field.onChange(fieldName)}
          value={field.value}
          onFocus={() => setIsFocused(true)}
          onBlur={() => setIsFocused(false)}
          {...props}
        />
      </View>
    </View>
  );
};

export default Input;

const styles = StyleSheet.create({
  group: {
    marginBottom: 10,
  },
  label: {
    marginBottom: 5,
  },
  input: {
    borderRadius: 5,
    borderWidth: StyleSheet.hairlineWidth,
    borderColor: theme.BORDER_COLOR_BUTTON,
    padding: 5,
  },
  errorInput: {
    borderColor: 'red',
  },
  focusedInput: {
    borderColor: theme.MAIN_COLOR,
  },
  error: {
    fontStyle: 'italic',
    color: 'red',
  },
});
