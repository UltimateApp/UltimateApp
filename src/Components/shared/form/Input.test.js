import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import { Formik } from 'formik';
import * as Yup from 'yup';
import { TouchableOpacity } from 'react-native';

import Input from './Input';

describe('Input', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Formik initialValues={{ title: 'My title' }}>
          <Input fieldName="title" label="Title" />
        </Formik>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when required', () => {
    const tree = renderer
      .create(
        <Formik initialValues={{ title: 'My title' }}>
          <Input fieldName="title" label="Title" required />
        </Formik>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when multiline', () => {
    const tree = renderer
      .create(
        <Formik initialValues={{ title: 'My title' }}>
          <Input fieldName="title" label="Title" multiline />
        </Formik>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when there is an error', async () => {
    const { toJSON, getByTestId, getByDisplayValue, getByText } = render(
      <Formik
        initialValues={{ title: 'My title' }}
        validationSchema={Yup.object({ title: Yup.string().required().min(1000, 'Invalid title') })}
      >
        {({ handleSubmit }) => (
          <>
            <Input fieldName="title" label="Title" />
            <TouchableOpacity onPress={handleSubmit} testID="submit" />
          </>
        )}
      </Formik>,
    );

    expect(getByDisplayValue('My title')).toBeDefined();

    await waitFor(() => {
      fireEvent.changeText(getByTestId('input'), 'new title');
    });

    expect(getByDisplayValue('new title')).toBeDefined();

    await waitFor(() => {
      fireEvent.press(getByTestId('submit'));
    });

    expect(getByText('Invalid title')).not.toBeNull();
    expect(toJSON()).toMatchSnapshot();
  });
});
