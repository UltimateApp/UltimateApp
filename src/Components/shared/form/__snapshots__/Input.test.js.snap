// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders correctly 1`] = `
<View
  style={
    Object {
      "marginBottom": 10,
    }
  }
>
  <Text
    style={
      Object {
        "marginBottom": 5,
      }
    }
  >
    Title
  </Text>
  <View
    style={
      Array [
        Object {
          "borderColor": "#e5e5e5",
          "borderRadius": 5,
          "borderWidth": 0.5,
          "padding": 5,
        },
        undefined,
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      rejectResponderTermination={true}
      underlineColorAndroid="transparent"
      value="Muy title"
    />
  </View>
</View>
`;
